File: backend/models/exam.py
from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, EmailStr, computed_field


class AnswerSubmission(BaseModel):
    question_id: int
    answer: Literal["A", "B", "C", "D"]


class ExamSubmission(BaseModel):
    email: EmailStr
    dni: str
    name: str
    surname: str
    exam_id: int
    answers: List[AnswerSubmission]
    accepts_marketing: bool = False


class ExamUserBase(BaseModel):
    name: str
    surname: str
    email: EmailStr | None = None
    dni: str
    accepts_marketing: bool = False


class ExamUserOut(ExamUserBase):
    id: int

    class Config:
        from_attributes = True


class ExamOut(BaseModel):
    id: int
    name: str
    score: Optional[float] = None
    percentile: Optional[float] = None
    show_score: bool = False
    show_percentile: bool = False
    show_score_full: bool = False
    validated_tribunal: bool = False

    class Config:
        from_attributes = True


class QuestionStubOut(BaseModel):
    id: int
    name: int
    is_active: bool
    is_cancelled: bool = False

    class Config:
        from_attributes = True


class QuestionCreate(BaseModel):
    correct_option: str  # 'A', 'B', 'C' o 'D'
    is_active: bool = True
    is_cancelled: bool = False
    name: Optional[int] = None


class ExamCreateWithQuestions(BaseModel):
    name: str
    is_active: Optional[bool] = None
    show_score: Optional[bool] = None
    show_percentile: Optional[bool] = None
    show_score_full: Optional[bool] = None
    validated_tribunal: Optional[bool] = None
    questions: List[QuestionCreate]


class QuestionEdit(BaseModel):
    id: Optional[int] = None
    correct_option: str
    is_active: bool = True
    is_cancelled: bool = False
    name: Optional[int] = None


class ExamEdit(BaseModel):
    name: Optional[str] = None
    is_active: Optional[bool] = None
    show_score: Optional[bool] = None
    show_percentile: Optional[bool] = None
    show_score_full: Optional[bool] = None
    validated_tribunal: Optional[bool] = None
    questions: List[QuestionEdit]


class AnswerReview(BaseModel):
    question_id: int
    question_label: Optional[int] = None
    selected_option: Optional[str] = None
    correct_option: Optional[str] = None
    is_correct: bool = False

    class Config:
        from_attributes = True


class AnswerDetail(BaseModel):
    id: int
    question_id: int
    answer: Literal["A", "B", "C", "D"]

    class Config:
        from_attributes = True


class UserExamSubmission(BaseModel):
    id: int
    exam_id: int
    user: ExamUserOut
    score: Optional[float] = None
    percentile: Optional[float] = None
    submitted_at: datetime

    class Config:
        from_attributes = True

    @computed_field
    def email(self) -> EmailStr | None:
        return self.user.email

    @computed_field
    def dni(self) -> str:
        return self.user.dni

    @computed_field
    def name(self) -> str:
        return self.user.name

    @computed_field
    def surname(self) -> str:
        return self.user.surname

    @computed_field
    def accepts_marketing(self) -> bool:
        return bool(getattr(self.user, "accepts_marketing", False))


class AdminSubmissionOut(UserExamSubmission):
    answers: List[AnswerDetail]


class AdminSubmissionUpdate(BaseModel):
    name: str
    surname: str
    email: EmailStr
    dni: str
    answers: List[AnswerSubmission]


class SubmissionCheckRequest(BaseModel):
    dni: str
    email: EmailStr
    exam_id: int


class SubmissionCheckResponse(BaseModel):
    score: Optional[float] = None
    percentile: float | None = None
    position: Optional[int] = None
    total_submissions: Optional[int] = None
    correct_answers: Optional[int] = None
    total_questions: Optional[int] = None
    message: Optional[str] = None
    answers_review: Optional[List[AnswerReview]] = None


class ExamOfficialResultOut(BaseModel):
    id: int
    exam_id: int
    user: ExamUserOut | None = None
    dni_masked: str
    apellido_1: str
    apellido_2: Optional[str]
    nombre: str
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend-exam/src/types/exam.ts
export type Exam = {
  id: number;
  name: string;
  slug: string;
  score?: number | null;
  percentile?: number | null;
  show_score?: boolean;
  show_percentile?: boolean;
  show_score_full?: boolean;
  validated_tribunal?: boolean;
  subtracts_points?: boolean;
  penalty_value?: number;
  max_score?: number;
  scoring_mode?: 'legacy' | 'absolute';
  points_per_correct?: number;
  points_per_wrong?: number;
  secondary_max_scores?: string;
  passing_criteria_type?: string;
  passing_criteria_value?: number | null;
  exam_weight?: number;
  max_merits?: number;
};
export type ExamOut = {
  score?: number | null;
  percentile?: number | null;
  position?: number | null;
  total_submissions?: number | null;
  correct_answers?: number | null;
  incorrect_answers?: number | null;
  not_answered?: number | null;
  total_questions?: number | null;
  message?: string | null;
  answers_review?: AnswerReview[] | null;
  max_score?: number | null;
  secondary_max_scores?: string | null;
  is_passed?: boolean | null;
  can_edit_merits?: boolean;
  max_merits?: number | null;
  merits?: number | null;
  weighted_score?: number | null;
  exam_weight?: number | null;
  merits_position?: number | null;
  merits_total?: number | null;
  passed_count?: number | null;
};

export type Question = {
  id: number;
  name: number;
  label?: string | null;
  correct_option?: string;
  is_active: boolean;
  is_cancelled?: boolean;
};

export type Answer = {
  question_id: number;
  answer: string;
}


export type ExamUser = {
  id: number;
  name: string;
  surname: string;
  email?: string | null;
  dni: string;
  moodle_id?: number | null;
  accepts_marketing?: boolean;
};

export type AdminSubmission = {
  id: number;
  exam_id: number;
  user?: ExamUser | null;
  email?: string | null;
  dni: string;
  name: string;
  surname: string;
  score?: number | null;
  merits?: number | null;
  percentile?: number | null;
  submitted_at: string;
  answers_data?: Record<string, string> | null;
  accepts_marketing?: boolean | null;
  selected_result_type?: string | null;
};

export type AdminSubmissionsResponse = {
  submissions: AdminSubmission[];
  total_submissions: number;
  average_score: number | null;
  stats_included: boolean;
};

export type SyncMoodleUsersResponse = {
  checked: number;
  synced: number;
  failed: number;
};

export type ExamSubmissionPayload = {
  email: string;
  dni: string;
  name: string;
  surname: string;
  exam_id: number;
  answers: Answer[];
  merits?: number;
  accepts_marketing: boolean;
  eligibility_confirmed?: boolean;
  result_type: string;
}

export type SubmissionUpdatePayload = {
  email: string;
  dni: string;
  name: string;
  surname: string;
  answers: Answer[];
  merits?: number | null;
};

export type ExamQuestionsResponse = {
  exam_name: string;
  questions: Question[];
};

export type OfficialResultCheckPayload = {
  exam_id: number;
  name: string;
  surname: string;
  dni: string;
};

export type OfficialResultCheckResponse = {
  match: boolean;
};

export type QuestionCreate = {
  id?: number;
  name?: number;
  label?: string | null;
  correct_option: string;
  is_active?: boolean;
  is_cancelled?: boolean;
};

export type ExamCreateWithQuestions = {
  name: string;
  is_active?: boolean;
  show_score?: boolean;
  show_percentile?: boolean;
  show_score_full?: boolean;
  validated_tribunal?: boolean;
  subtracts_points?: boolean;
  penalty_value?: number;
  max_score?: number;
  scoring_mode?: 'legacy' | 'absolute';
  points_per_correct?: number;
  points_per_wrong?: number;
  secondary_max_scores?: string;
  passing_criteria_type?: string;
  passing_criteria_value?: number | null;
  exam_weight?: number;
  max_merits?: number;
  display_exam_weight?: number | null;
  skip_weights?: boolean;
  questions: QuestionCreate[];
};

export type QuestionEdit = {
  id?: number;
  name?: number;
  label?: string | null;
  correct_option: string;
  is_active?: boolean;
  is_cancelled?: boolean;
};

export type ExamEdit = {
  id?: number;
  name?: string;
  is_active?: boolean;
  show_score?: boolean;
  show_percentile?: boolean;
  show_score_full?: boolean;
  validated_tribunal?: boolean;
  subtracts_points?: boolean;
  penalty_value?: number;
  max_score?: number;
  scoring_mode?: 'legacy' | 'absolute';
  points_per_correct?: number;
  points_per_wrong?: number;
  secondary_max_scores?: string;
  passing_criteria_type?: string;
  passing_criteria_value?: number | null;
  exam_weight?: number;
  max_merits?: number;
  display_exam_weight?: number | null;
  clear_display_weight?: boolean;
  skip_weights?: boolean;
  associated_exam_ids?: number[];
  questions: QuestionEdit[];
};

export type AnswerReview = {
  question_id: number;
  question_label?: string | null;
  selected_option?: string | null;
  correct_option?: string | null;
  is_correct: boolean;
  has_feedback_video?: boolean;
};

export type UserSubmissionCheck = {
  email: string;
  dni: string;
  exam_id: number;
};

export type ExamOfficialResult = {
  id: number;
  exam_id: number;
  user?: ExamUser | null;
  dni_masked: string;
  apellido_1: string;
  apellido_2?: string | null;
  nombre: string;
  result_type: string;
  score?: number | null;
  merits?: number | null;
  created_at: string;
};

export type CreateOfficialResultPayload = {
  dni: string;
  apellido_1: string;
  apellido_2?: string | null;
  nombre: string;
  result_type: string;
  score?: number | null;
  merits?: number | null;
};

export type EditOfficialResultPayload = {
  dni?: string;
  apellido_1?: string;
  apellido_2?: string | null;
  nombre?: string;
  result_type?: string;
  score?: number | null;
  merits?: number | null;
};


export type AdminOfficialResultsResponse = {
  results: ExamOfficialResult[];
  total: number;
};

export type ImportOfficialResultsSummary = {
  exam_id: number;
  total_rows: number;
  imported_results: number;
  created_users: number;
  updated_users: number;
};

export type SortableHeaderProps = {
  label: string;
  sortKey: 'dni' | 'nombre' | 'apellidos' | 'usuario' | 'creado';
  activeKey: 'dni' | 'nombre' | 'apellidos' | 'usuario' | 'creado';
  direction: 'asc' | 'desc';
  onSort: (key: 'dni' | 'nombre' | 'apellidos' | 'usuario' | 'creado') => void;
};

export type UpdateMeritsPayload = {
  email: string;
  dni: string;
  exam_id: number;
  merits: number | null;
};

export type UpdateMeritsResponse = {
  message: string;
  merits: number | null;
  weighted_score?: number | null;
  merits_position?: number | null;
  merits_total?: number | null;
};

export type ExamUiState = {
  checking: boolean;
  hasPreviousSubmission: boolean;
  score: number | null;
  percentile: number | null;
  position: number | null;
  totalSubmissions: number | null;
  correctAnswers: number | null;
  incorrectAnswers: number | null;
  notAnswered: number | null;
  totalQuestions: number | null;
  submissionMessage: string | null;
  resultError: string | null;
  answersReview: AnswerReview[] | null;
  maxScore: number | null;
  secondaryMaxScores: string | null;
  isPassed: boolean | null;
  canEditMerits: boolean;
  maxMerits: number | null;
  merits: number | null;
  weightedScore: number | null;
  examWeight: number | null;
  meritsPosition: number | null;
  meritsTotal: number | null;
  passedCount: number | null;
};

export type ExamResultPayload = {
  score: number | null;
  percentile: number | null;
  position: number | null;
  totalSubmissions: number | null;
  correctAnswers: number | null;
  incorrectAnswers: number | null;
  notAnswered: number | null;
  totalQuestions: number | null;
  message: string;
  answersReview: AnswerReview[] | null;
  max_score?: number | null;
  secondary_max_scores?: string | null;
  isPassed: boolean | null;
  canEditMerits: boolean;
  maxMerits: number | null;
  merits: number | null;
  weightedScore: number | null;
  examWeight: number | null;
  meritsPosition: number | null;
  meritsTotal: number | null;
  passedCount: number | null;
};

export type ExamUiAction =
  | { type: 'RESET' }
  | { type: 'CHECK_START' }
  | { type: 'CHECK_SUCCESS'; payload: ExamResultPayload }
  | { type: 'CHECK_ERROR'; payload?: string | null }
  | { type: 'SUBMIT_SUCCESS'; payload: ExamResultPayload }
  | { type: 'SUBMIT_ERROR'; payload?: string | null };


